default/python/AI/buildings.py

```diff
@@ -39,12 +46,28 @@ def available(self, eid: EmpireId = THIS_EMPIRE) -> bool:
             empire_object = fo.getEmpire(eid)
         return empire_object.buildingTypeAvailable(self.value)
 
+    def get_opinions(self) -> Opinion:
+        return get_planet_opinion(self.value)
+
     def queued_in(self) -> List[PlanetId]:
         """
         Return list of planet ids where this building is queued.
         """
         return [element.locationID for element in fo.getEmpire().productionQueue if (element.name == self.value)]
 
+    def built_at(self, include_queued_buildings: bool = False) -> Set[PlanetId]:
+        ret = _get_building_locations()[self.value].planets
+        if include_queued_buildings:
+            ret.update(set(self.queued_in()))
+        return ret
+
+    def built_at_sys(self, include_queued_buildings: bool = False) -> Set[SystemId]:
+        ret = _get_building_locations()[self.value].systems
+        if include_queued_buildings:
+            universe = fo.getUniverse()
+            ret.update(set(universe.getPlanet(pid).systemID for pid in self.queued_in()))
+        return ret
+
```

Comment:
```suggestion
        return {universe.getPlanet(pid).systemID for pid in self.built_at(include_queued_buildings=include_queued_buildings)}
```

Reply:
Well, actually I think I'll revert this, _get_building_locations()[self.value].systems returns a cached set, using it should be faster then building a new set with every call, shouldn't it?

Reply:
> Well, actually I think I'll revert this, _get_building_locations()[self.value].systems returns a cached set, using it should be faster then building a new set with every call, shouldn't it?

Cashe is not free, set creation is fast. Keep it simple.    We could add caching later, let's focus on the logic.

Reply:
I've done the caching already and if we implement your suggestion, we'd use the same cached function indirectly via built_at.